src/pg_copy.c

```diff
@@ -0,0 +1,552 @@
+			}
+
+			while (nextRowFound)
+			{
+				MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+				nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+				MemoryContextSwitchTo(oldContext);
+					
+				/* write the row to the shard */
+				if (nextRowFound)
+				{	
+					if (columnNulls[partitionColumn->varattno]) 
+					{ 
+						ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+										errmsg("cannot copy row with NULL value "
+											   "in partition column")));
+					} 
+					rightConst->constvalue = columnValues[partitionColumn->varattno];
+					prunedList = PruneShardList(tableId, whereClauseList, shardIntervalList);
+					
+					foreach(shardIntervalCell, prunedList)
+					{
+						ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+						ShardId shardId = shardInterval->id;
+						bool found = false;
+						
+						shardConn = (ShardConnections*)hash_search(shardToConn, &shardInterval->id, HASH_ENTER, &found);
+						if (!found) { 
+							List *finalizedPlacementList = NULL;
+                            int nPlacements = 0;
+
+							finalizedPlacementList = LoadFinalizedShardPlacementList(shardId); 
+                            nPlacements = list_length(finalizedPlacementList);
+                            shardConn->conn = (PGconn**)palloc0(sizeof(PGconn*)*nPlacements);
+							shardConn->nReplicas = nPlacements;
+							shardConn->status = (bool*)palloc0(nPlacements*sizeof(bool));
+                            nPlacements = 0;
+
+							foreach(taskPlacementCell, finalizedPlacementList)
+							{
+								ShardPlacement *taskPlacement = (ShardPlacement *) lfirst(taskPlacementCell);
+								char *nodeName = taskPlacement->nodeName;
+								char nodePort[16];
+								PGconn* conn = NULL;
+								sprintf(nodePort, "%d", taskPlacement->nodePort);	 
+								conn = ConnectToNode(nodeName, nodePort);
+								if (conn != NULL)										 
+								{
+                                    char const* copy = ConstructCopyStatement(copyStatement, shardId);
+									shardConn->conn[nPlacements++] = conn;
+									/*
+									 * New connection: start transaction with copy command on it.
+									 * Append shard id to table name.
+									 */
+									if (!tmgr->Begin(conn) ||
+										!PgShardExecute(conn, PGRES_COPY_IN, copy))
+									{
+										elog(ERROR, "Failed to start '%s' on node %s:%s", copy, nodeName, nodePort);
+									}
+								} else { 
+									elog(ERROR, "Failed to connect to node %s:%s", nodeName, nodePort);
+								}
+							}
+						}
+						/* TODO: handle binary format */
+						lineBuf = CopyGetLineBuf(copyState);
+						lineBuf->data[lineBuf->len++] = '\n'; /* there was already new line in the buffer, but it was truncated: no need to heck available space */
+						/* Replicate row to all shards */
+						for (i = 0; i < shardConn->nReplicas; i++) { 
+							PQputCopyData(shardConn->conn[i], lineBuf->data, lineBuf->len);
```

Comment:
What if there's an error here? Does it always fall into PG_CATCH()? The documentation says it just returns -1.